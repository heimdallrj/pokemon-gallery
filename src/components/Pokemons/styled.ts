import styled from 'styled-components';

export const Conatiner = styled.ul`
  list-style: none;
  display: flex;
  gap: 15px;
  justify-content: flex-start;
  align-items: flex-start;
  flex-wrap: wrap;
`;

export const Pokemon = styled.li`
  padding: 15px;
  border: 1px solid #ccc;
  color: #5757bb;
  cursor: pointer;
  display: flex;
  align-items: center;

  &:hover {
    background-color: #ccc;
  }
`;
