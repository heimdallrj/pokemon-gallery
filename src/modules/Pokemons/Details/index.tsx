import { useEffect, useState } from 'react';

import { fetchOne as apiFetchPokemon } from 'api/pokemon';

import Pokemon from 'components/Pokemon';

type Props = {
  name?: string;
};

export default function PokemonDetails({ name }: Props) {
  const [isValidPokemon, setIsValidPokemon] = useState(!name);
  const [avatar, setAvatar] = useState();
  const [weight, setWeight] = useState(null);
  const [abilities, setAbilities] = useState([]);
  const [moves, setMoves] = useState([]);

  useEffect(() => {
    if (!name) return;

    (async () => {
      const _pokemon: any = await apiFetchPokemon(name);
      setIsValidPokemon(!!_pokemon);

      // weight
      setWeight(_pokemon.weight || null);

      // avatar
      const _avatar = _pokemon.sprites.front_default;
      setAvatar(_avatar);

      //abilities
      const _abilities = _pokemon.abilities
        .map((obj: any) => obj.ability)
        .map((obj: any) => obj.name);
      setAbilities(_abilities);

      // moves
      const _moves = _pokemon.moves
        .map((obj: any) => obj.move)
        .map((obj: any) => obj.name);
      setMoves(_moves);
    })();
  }, [name]);

  if (!name || !isValidPokemon) return <div>No Pok√©mon in the area.</div>;

  return (
    <Pokemon
      name={name}
      avatar={avatar}
      weight={weight}
      abilities={abilities}
      moves={moves}
    />
  );
}
