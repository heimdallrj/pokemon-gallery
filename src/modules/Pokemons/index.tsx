import { useEffect, useState } from 'react';

import usePokemons from 'hooks/usePokemons';

import PokemonsList from 'components/Pokemons';
import Spinner from 'components/Spinner';

import {
  Container,
  Error,
  SpinnerWrapper,
  LoadMoreWrapper,
  LoadMoreBtn,
} from './styled';

const LIMIT = 50;
const lsFavKey = '__favourites';

type Props = {
  onSelect: (name: string) => void;
};

export default function Pokemons({ onSelect }: Props) {
  const [offset, setOffset] = useState(0);
  const [favourites, setFavourites] = useState<string[]>([]);

  const {
    data: pokemons,
    loadMore,
    isLoading,
    error,
  } = usePokemons({ offset, limit: LIMIT });

  useEffect(() => {
    // Fetch favourites from localStorage
    const _favourites = window.localStorage.getItem(lsFavKey);
    if (_favourites) {
      const _favouritesParsed = JSON.parse(_favourites);
      setFavourites(_favouritesParsed);
    }
  }, []);

  const handleFavourite = (name: string) => {
    if (!favourites.includes(name)) {
      setFavourites((prevState: string[]) => [...prevState, name]);
    } else {
      // un-Like
      const _favourites = favourites.filter((f) => f !== name);
      setFavourites(_favourites);
    }

    // Sync with localStorage
    window.localStorage.setItem(lsFavKey, JSON.stringify(favourites));
  };

  const handleLoadMore = () => {
    setOffset(offset + LIMIT);
  };

  return (
    <Container>
      <PokemonsList
        isLoading={isLoading}
        pokemons={pokemons}
        favourites={favourites}
        onFavourite={handleFavourite}
        onSelect={onSelect}
      />

      {error && <Error>Oops! something went wrong</Error>}

      {isLoading && (
        <SpinnerWrapper>
          <Spinner />
        </SpinnerWrapper>
      )}

      {loadMore && (
        <LoadMoreWrapper>
          <LoadMoreBtn onClick={handleLoadMore}>
            Click here to load more &#8595;
          </LoadMoreBtn>
        </LoadMoreWrapper>
      )}
    </Container>
  );
}
