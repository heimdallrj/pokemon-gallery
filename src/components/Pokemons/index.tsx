import Favourite from 'components/Favourite';
import { IPokemon } from 'types';

import { Conatiner, Pokemon } from './styled';

type Props = {
  isLoading?: boolean;
  pokemons: IPokemon[];
  favourites: string[];
  onFavourite: (name: string) => void;
  onSelect: (name: string) => void;
};

export default function Pokemons({
  isLoading = true,
  pokemons = [],
  favourites = [],
  onFavourite,
  onSelect,
}: Props) {
  return (
    <>
      {!isLoading && pokemons.length === 0 && (
        <div>No Pok√©mon in the area.</div>
      )}
      <Conatiner>
        {pokemons.map(({ name }: { name: string }) => (
          <Pokemon key={name}>
            <Favourite
              liked={favourites.includes(name)}
              onClick={() => onFavourite(name)}
            />
            <p onClick={() => onSelect(name)}>{name}</p>
          </Pokemon>
        ))}
      </Conatiner>
    </>
  );
}
