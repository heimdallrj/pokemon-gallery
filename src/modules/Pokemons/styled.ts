import styled from 'styled-components';

export const Container = styled.div`
  width: 100;
`;

export const Error = styled.div`
  color: red;
  text-align: center;
  padding: 30px 0 0;
`;

export const SpinnerWrapper = styled.div`
  display: flex;
  justify-content: center;
  padding: 30px 0 0;
`;

export const LoadMoreWrapper = styled.div`
  display: flex;
  justify-content: center;
`;

export const LoadMoreBtn = styled.button`
  margin: 30px 0;
  outline: none;
  color: blue;
  background: transparent;
  font-size: 14px;
  cursor: pointer;
`;
