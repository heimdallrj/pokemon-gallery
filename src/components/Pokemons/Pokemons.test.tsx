import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';

import Pokemons from './index';

const pokemons: any = [
  {
    name: 'venonat',
    url: 'url1',
  },
  {
    name: 'venusaur',
    url: 'url2',
  },
];

describe('Components - Pokemons', () => {
  test('renders when no data has provided', () => {
    const component = (
      <Pokemons
        isLoading={false}
        pokemons={[]}
        favourites={[]}
        onFavourite={() => {}}
        onSelect={() => {}}
      />
    );
    render(component);

    expect(screen.getByText(/No PokÃ©mon in the area./i)).toBeInTheDocument();

    // Snapshot
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test.only('renders when data has provided', async () => {
    const handleFavourite = jest.fn();
    const handleSelect = jest.fn();

    const component = (
      <Pokemons
        isLoading={false}
        pokemons={pokemons}
        favourites={[]}
        onFavourite={handleFavourite}
        onSelect={handleSelect}
      />
    );
    render(component);

    expect(screen.getByText(/venonat/i)).toBeInTheDocument();
    expect(screen.getByText(/venusaur/i)).toBeInTheDocument();

    // screen.debug();

    fireEvent.click(screen.getByText(/venonat/i));
    expect(handleSelect).toHaveBeenCalledTimes(1);

    // eslint-disable-next-line testing-library/no-node-access
    const container: any = document.querySelector('svg')?.firstChild;
    fireEvent.click(container);
    expect(handleFavourite).toHaveBeenCalledTimes(1);

    // Snapshot
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
