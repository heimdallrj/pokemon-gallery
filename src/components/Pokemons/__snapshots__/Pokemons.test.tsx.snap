// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components - Pokemons renders when data has provided 1`] = `
<ul
  className="sc-dkPtRN evmAl"
>
  <li
    className="sc-hKwDye bqwgDE"
  >
    <svg
      className="sc-bdvvtL jbVkAX"
      fill="none"
      onClick={[Function]}
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </svg>
    <p
      onClick={[Function]}
    >
      venonat
    </p>
  </li>
  <li
    className="sc-hKwDye bqwgDE"
  >
    <svg
      className="sc-bdvvtL jbVkAX"
      fill="none"
      onClick={[Function]}
      stroke="currentColor"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth={2}
      />
    </svg>
    <p
      onClick={[Function]}
    >
      venusaur
    </p>
  </li>
</ul>
`;

exports[`Components - Pokemons renders when no data has provided 1`] = `
Array [
  <div>
    No Pok√©mon in the area.
  </div>,
  <ul
    className="sc-dkPtRN evmAl"
  />,
]
`;
